File: src/state/logger.ts
import * as fs from "node:fs";
import * as path from "node:path";
import { shouldRedact } from "./secretRedactor";

export type LogLevel = "debug" | "info" | "warn" | "error";

export interface LogEntry {
  ts: string;
  level: LogLevel;
  sessionId: string;
  iteration?: number;
  category: string;
  message: string;
  data?: Record<string, unknown>;
}

export class StructuredLogger {
  private readonly logPath: string;
  private buffer: string[] = [];
  private flushTimer: ReturnType<typeof setTimeout> | null = null;
  private readonly maxBuffer = 50;
  private readonly flushDelayMs = 100;

  constructor(logDir: string, sessionId: string) {
    this.logPath = path.join(logDir, `${sessionId}.ndjson`);
  }

  log(entry: LogEntry): void {
    const redacted = this.redactEntry(entry);
    this.buffer.push(JSON.stringify(redacted));
    if (this.buffer.length >= this.maxBuffer) {
      this.flush();
    } else if (!this.flushTimer) {
      this.flushTimer = setTimeout(() => this.flush(), this.flushDelayMs);
    }
  }

  close(): void {
    if (this.flushTimer) {
      clearTimeout(this.flushTimer);
      this.flushTimer = null;
    }
    this.flush();
  }

  private flush(): void {
    if (this.buffer.length === 0) return;
    const chunk = `${this.buffer.join("\n")}\n`;
    this.buffer = [];
    try {
      fs.mkdirSync(path.dirname(this.logPath), { recursive: true });
      fs.appendFileSync(this.logPath, chunk, { encoding: "utf-8" });
    } catch {
      // Best-effort: log writes must never crash the agent loop.
    }
  }

  private redactEntry(entry: LogEntry): LogEntry {
    if (!entry.data) return entry;
    return { ...entry, data: redactObject(entry.data) };
  }
}

function redactObject(obj: Record<string, unknown>): Record<string, unknown> {
  const out: Record<string, unknown> = {};
  for (const [k, v] of Object.entries(obj)) {
    if (shouldRedact(k)) {
      out[k] = "***";
    } else if (v !== null && typeof v === "object" && !Array.isArray(v)) {
      out[k] = redactObject(v as Record<string, unknown>);
    } else {
      out[k] = v;
    }
  }
  return out;
}

File: src/infra_agent/state/logger.py
"""Structured NDJSON logger with secret redaction.
"""

import json
import os
import threading
from pathlib import Path
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field

from infra_agent.state.secret_redactor import redact_text, redact_value_deep

LogLevel = Literal["debug", "info", "warn", "error"]


class LogEntry(BaseModel):
    """One log row. ``message`` and ``data`` are redacted by the logger."""

    model_config = ConfigDict(extra="forbid", populate_by_name=True)
    ts: str
    level: LogLevel
    session_id: str = Field(alias="sessionId")
    iteration: int | None = None
    category: str
    message: str
    data: dict[str, Any] | None = None


class StructuredLogger:
    """Buffered, best-effort NDJSON logger per session."""

    def __init__(self, log_dir: str | os.PathLike[str], session_id: str) -> None:
        self._log_path = Path(log_dir) / f"{session_id}.ndjson"
        self._buffer: list[str] = []
        self._max_buffer = 50
        self._flush_delay_ms = 100
        self._timer: threading.Timer | None = None
        self._lock = threading.Lock()

    def log(self, entry: LogEntry) -> None:
        """Buffer a redacted log entry; flush synchronously if threshold hit."""
        redacted = self._redact_entry(entry)
        should_flush = False
        with self._lock:
            self._buffer.append(json.dumps(redacted.model_dump(by_alias=True)))
            if len(self._buffer) >= self._max_buffer:
                should_flush = True
            elif self._timer is None:
                self._timer = threading.Timer(self._flush_delay_ms / 1000.0, self._flush)
                self._timer.start()
        if should_flush:
            self._flush()

    def close(self) -> None:
        """Cancel pending timer and flush remaining entries."""
        with self._lock:
            if self._timer is not None:
                self._timer.cancel()
                self._timer = None
        self._flush()

    def _redact_entry(self, entry: LogEntry) -> LogEntry:
        updates: dict[str, Any] = {"message": redact_text(entry.message)}
        if entry.data is not None:
            redacted = redact_value_deep(entry.data)
            if isinstance(redacted, dict):
                updates["data"] = redacted
        return entry.model_copy(update=updates)

    def _flush(self) -> None:
        with self._lock:
            if not self._buffer:
                return
            chunk = "\n".join(self._buffer) + "\n"
            self._buffer = []
        try:
            self._log_path.parent.mkdir(parents=True, exist_ok=True)
            with self._log_path.open("a", encoding="utf-8") as f:
                f.write(chunk)
        except Exception:  # noqa: BLE001
            # Best-effort: log writes must never crash the agent loop.
            pass


__all__ = ["LogEntry", "LogLevel", "StructuredLogger"]